src/commands/shell.ts

```diff
@@ -0,0 +1,120 @@
+import { getRefreshToken } from "../auth";
+import { requireAuth } from "../requireAuth";
+import { FirebaseError } from "../error";
+import { createContext, runInContext, Context, Script } from "vm";
+import { readFileSync, existsSync } from "fs";
+import fetch, { Response } from "node-fetch";
+import { DocumentSnapshot, Firestore, QuerySnapshot } from "@google-cloud/firestore";
+import api = require("../api");
+
+async function runScript(sandbox: Context, scriptPath: string) {
+  if (!existsSync(scriptPath)) {
+    throw new FirebaseError(`Script file ${bold(scriptPath)} does not exist.`);
+  }
+
+  try {
+    const script = new Script(readFileSync(scriptPath, { encoding: "utf8" }), {
+      filename: scriptPath,
+    });
+    return await Promise.resolve(script.runInContext(sandbox));
+  } catch (e) {
+    // show the full stacktrace for the user's custom script
+    throw new FirebaseError(`Error running script: ${e.stack}`);
+  }
+}
+
+function runRepl(sandbox: Context) {
+  async function shellEval(command: string, context: any, filename: string, callback: any) {
+    try {
+      let result = await Promise.resolve(
+        // Wrap in an async function to allow top-level await.
+        runInContext(`(async function(){ return ${command}; })()`, sandbox)
+      );
+
+      if (result instanceof QuerySnapshot) {
+        result = result.docs.map((doc) => simpleDocSnapshot(doc));
+      } else if (result instanceof DocumentSnapshot) {
+        result = simpleDocSnapshot(result);
+      } else if (typeof result.val === "function") {
+        result = result.val();
+      } else if (result instanceof Response) {
+        if (result.headers.get("content-type")?.includes("application/json")) {
+          result = await result.json();
+        } else {
+          result = await result.text();
+        }
+      }
+
+      callback(null, result);
+    } catch (e) {
+      callback(e);
+    }
+  }
+
+  const replServer = start({
+    prompt: `${yellow("firebase")}${red(">")} `,
+    eval: shellEval,
+    // writer: replWriter,
+  });
+
+  return new Promise((resolve) => {
+    replServer.on("exit", resolve);
+    process.on("SIGINT", resolve);
+  });
+}
+
+function simpleDocSnapshot(snap: DocumentSnapshot) {
+  return Object.assign({ __id__: snap.id }, snap.data());
+}
+
+module.exports = new Command("shell [script_path]")
+  .description("an interactive shell for admin project access")
+  .before(requireAuth)
+  .action(async (scriptPath: string, options: any) => {
+    const sdkConfig = await getFirebaseConfig(options);
+    admin.initializeApp(Object.assign({}, sdkConfig, { credential: { getAccessToken } }));
+    const accessToken = ((await getAccessToken()) as any).access_token;
+
+    const firestoreConfig = {
+      projectId: sdkConfig.projectId,
+      credentials: {
+        type: "authorized_user",
+        client_id: api.clientId,
+        client_secret: api.clientSecret,
+        refresh_token: getRefreshToken(),
+      } as any,
```

Comment:
The Firestore library doesn't have a type for this? I thought that one carried its own types. It'd be nice to not cast something as `any` if we can help it.

Reply:
It's a hack to allow user credentials, which is an undocumented but technically supported feature. See [this issue](https://github.com/googleapis/nodejs-firestore/issues/973) which I will add in a comment.